src/main/java/org/opensearch/ad/rest/handler/ModelValidationActionHandler.java

```diff
@@ -0,0 +1,718 @@
+        Client client,
+        ActionListener<ValidateAnomalyDetectorResponse> listener,
+        AnomalyDetector anomalyDetector,
+        TimeValue requestTimeout,
+        NamedXContentRegistry xContentRegistry,
+        SearchFeatureDao searchFeatureDao,
+        String validationType,
+        Clock clock
+    ) {
+        this.clusterService = clusterService;
+        this.client = client;
+        this.listener = listener;
+        this.anomalyDetector = anomalyDetector;
+        this.requestTimeout = requestTimeout;
+        this.xContentRegistry = xContentRegistry;
+        this.searchFeatureDao = searchFeatureDao;
+        this.validationType = validationType;
+        this.clock = clock;
+    }
+
+    // Need to first check if multi entity detector or not before doing any sort of validation.
+    // If detector is HCAD then we will find the top entity and treat as single entity for
+    // validation purposes
+    public void checkIfMultiEntityDetector() {
+        ActionListener<Map<String, Object>> recommendationListener = ActionListener
+            .wrap(topEntity -> getLatestDateForValidation(topEntity), exception -> {
+                listener.onFailure(exception);
+                logger.error("Failed to get top entity for categorical field", exception);
+            });
+        if (anomalyDetector.isMultientityDetector()) {
+            getTopEntity(recommendationListener);
+        } else {
+            recommendationListener.onResponse(Collections.emptyMap());
+        }
+    }
+
+    // For single category HCAD, this method uses bucket aggregation and sort to get the category field
+    // that have the highest document count in order to use that top entity for further validation
+    // For multi-category HCADs we use a composite aggregation to find the top fields for the entity
+    // with the highest doc count.
+    private void getTopEntity(ActionListener<Map<String, Object>> topEntityListener) {
+        AggregationBuilder bucketAggs;
+        Map<String, Object> topKeys = new HashMap<>();
+        if (anomalyDetector.getCategoryField().size() == 1) {
+            bucketAggs = AggregationBuilders
+                .terms(AGG_NAME_TOP)
+                .field(anomalyDetector.getCategoryField().get(0))
+                .order(BucketOrder.count(true));
+        } else {
+            bucketAggs = AggregationBuilders
+                .composite(
+                    AGG_NAME_TOP,
+                    anomalyDetector
+                        .getCategoryField()
+                        .stream()
+                        .map(f -> new TermsValuesSourceBuilder(f).field(f))
+                        .collect(Collectors.toList())
+                )
+                .size(1000)
+                .subAggregation(
+                    PipelineAggregatorBuilders
+                        .bucketSort("bucketSort", Collections.singletonList(new FieldSortBuilder("_count").order(SortOrder.DESC)))
+                        .size(1)
```

Comment:
It's understandable to only get the top entity combination to use when doing the sparsity checks. But just thinking, what if the data has one or two dominant/frequent entity combos, and many more that are less common, but still useful? For example, maybe an interval of 5 mins is above the threshold for the top 1 entity combo, but an interval of 10 mins is above the threshold for the top 20 entity combos, in which case a suggestion of 10 mins interval would be much more useful from the customer perspective. Is it easy enough to repeat this for the top x entity combos and try to derive a more general working interval suggestion? (this applies to single-category HCAD as well, just using multi-category as the example).

Reply:
I could implement this within the first check when all configurations are applied. It would mean running multiple intervals for each one of the top 10 entities and keeping track of the highest interval. The best case is when I find an interval for all top entities but I guess if there is only an interval found for 5 of them, maybe it means the category fields should be relooked at or it could be good enough? On the check where I am adding the category field to the configurations I could repeat this as well but first goal could be only to initially do it for the first run with all configurations.

Reply:
Yeah it's impossible to know for sure what would be the ideal number of running entities. Roughly, my idea would be
1. get top x entities (if < y then just optimize for single entity)
2. find a baseline interval that works for single top entity (what you're already doing)
3. check if that works for top z entities - if not, increase it a few more times to see if it can go over the threshold for top z entities - if not, then stick with baseline found in step 2

where x = num top entities constant, y = num top entities threshold, z = num top entities working given the interval threshold

Reply:
^ Id consider this not a hard reqt for this release, just an idea.